crates/sui-framework/src/natives/crypto.rs

```diff
@@ -40,3 +40,38 @@ pub fn ecrecover(
         Err(_) => Ok(NativeResult::err(cost, INVALID_SIGNATURE)),
     }
 }
+
+pub fn bls12381_verify(
+    context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(ty_args.is_empty());
+    debug_assert!(args.len() == 3);
+
+    let msg = pop_arg!(args, Vec<u8>);
+    let public_key_bytes = pop_arg!(args, Vec<u8>);
+    let signature_bytes = pop_arg!(args, Vec<u8>);
+
+    // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/3868
+    let cost = native_gas(context.cost_table(), NativeCostIndex::EMIT_EVENT, 0);
+
+    let signature = match <narwhal_crypto::bls12381::BLS12381Signature as ToFromBytes>::from_bytes(
+        &signature_bytes,
+    ) {
+        Ok(signature) => signature,
+        Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
+    };
+
+    let public_key = match <narwhal_crypto::bls12381::BLS12381PublicKey as ToFromBytes>::from_bytes(
+        &public_key_bytes,
+    ) {
+        Ok(public_key) => public_key,
+        Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
+    };
+
+    match public_key.verify(&msg, &signature) {
```

Comment:
is this the raw message or the hashed message? maybe worth adding in the comment

Reply:
It is the raw message - does it have to be a hash?

Reply:
That is also fine to do a hash - I only need the hash anyways.